source/MaterialXGenShader/ShaderNode.h

```diff
@@ -7,83 +7,141 @@
 #include <MaterialXGenShader/TypeDesc.h>
 #include <MaterialXGenShader/ShaderNodeImpl.h>
 
+#include <memory>
 #include <set>
 
 namespace MaterialX
 {
 
 class ShaderNode;
+class ShaderPort;
 class ShaderInput;
 class ShaderOutput;
 
+using ShaderPortPtr = shared_ptr<class ShaderPort>;
 using ShaderInputPtr = shared_ptr<class ShaderInput>;
 using ShaderOutputPtr = shared_ptr<class ShaderOutput>;
 using ShaderNodePtr = shared_ptr<class ShaderNode>;
 using ShaderInputSet = std::set<ShaderInput*>;
 
 /// An input or output port on a ShaderNode
-class ShaderPort
+class ShaderPort : public std::enable_shared_from_this<ShaderPort>
 {
   public:
-    static const unsigned int VARIABLE_NOT_RENAMABLE = 1 << 0; // Variable should not be automatically named
+    /// Flags set on shader ports.
+    static const unsigned int EMITTED = 1 << 0;
 
-    /// Copy data from another port to this port
-    void copyData(const ShaderPort& other)
-    {
-        value = other.value;
-        path = other.path;
+    ShaderPort(ShaderNode* node, const TypeDesc* type, const string& name, ValuePtr value = nullptr);
 
-        if (ShaderPort::VARIABLE_NOT_RENAMABLE & other.flags)
```

Comment:
@feldstj I didn't quite understand why this flag was used. I removed this logic and all tests are passing, including color management. But let me know if I am missing something.

Reply:
Maybe we are missing a test? Because copy data won't be called any more for constants with CM because they aren't optimized out. So the existing tests won't use this. Not sure how to easily make a test that will use this though. Need some other node type that can be optimized out I guess.

Reply:
I'm not sure what the original problem was. Having a flag for disallowing variable renaming seems risky as that might result in name conflicts and code that won't compile. Maybe we could work around the original problem in another way, if the original problem still exists. We can take this offline.